### DISCUSSION TITLE
0065 XLS - 65d: Single Asset Vault

### DISCUSSION THREAD

**Discussion by Tapanito** [XRPLF/XRPL-Standards#192] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by xVet** [comment]
For the CUD operations, i wonder, why is the `PoolDelegate` the only account able to perform update operations and delete operations. Including updates to the PoolDelegateID.

I would expect that the AccountID creating the `Pool` is the owner and delegates to the `PoolDelegate` limited functionality.

**Reply by Tapanito** [reply]
Thanks @xVet !

The idea is that a `PoolDelegate` is the pool's owner, and thus, the `PoolDelegateID` is not changeable. However, now that I think about it, there is no "delegacy" in this case. 

Do you think it would be better to have a separate `PoolOwner` and `PoolDelegate` where the `PoolOwner` has all permissions of the Delegate, including deleting the pool and changing the PoolDelegate. Whereas the PoolDelegate may only perform protocol specific operations?

**Reply by rothexD** [reply]
Following the discussion we had in https://github.com/XRPLF/XRPL-Standards/discussions/190#discussioncomment-9118533

I would propose the following:

I think a `Pool` should be something generic with no direct owner and no delegate, a tokenized storage for a asset where we can branch of other types of `Pools`, a `Vault` is a type of `Pool` that has a owner and managed delegate. 

For example a number of XRP users can lock up XRP in a `Flash-Loan-Pool` and based on their LP tokens they can vote for a fee, users can then do a `Remit/Atomic` transaction to do some flash loan transaction chain. This `Flash-Loan-Pool` would have no owner or delegate as its management is done decentralized.

I guess it really depends on where you want to go with the `Pool` but with the current design doing a `Flash-Loan-Pool` would require you to set some pseudo accounts etc 

------
Even if we have a `PoolOwner` and `PoolDelegate` I think we would need functionality to change and transfer either of them. If a `PoolOwner` is bought by someone or goes into bankruptcy the `Pool` will need to be transferred and held by someone else.

A `Pool` might have a single debtor and when you think about car sales or similar usually they reserve the right to sell the loan-contract to a 3rd party at a later point of time

**Reply by mvadari** [reply]
In terms of terminology the `Pool`/`Vault` difference makes sense, but in terms of implementation, ledger objects don't have any inheritance structure so IMO we should just have `Pool`, which has optional `Owner`/`Delegate` fields.

**Reply by rothexD** [reply]
> In terms of terminology the `Pool`/`Vault` difference makes sense, but in terms of implementation, ledger objects don't have any inheritance structure so IMO we should just have `Pool`, which has optional `Owner`/`Delegate`

I don't think optional would be a good idea then.

Sounds interesting, I think we should consider standardized unambiguous language in some way and always say any type of `Pool` has no owner and any type of `Vault` has an owner or the other way around.

If we want to use `Pool` in both cases I propose renaming it to something like `Managed Single Asset Tokenized Pool` to show specifically that this pool is managed by someone

**Reply by Tapanito** [reply]
> For example a number of XRP users can lock up XRP in a Flash-Loan-Pool and based on their LP tokens they can vote for a fee, users can then do a Remit/Atomic transaction to do some flash loan transaction chain. This Flash-Loan-Pool would have no owner or delegate as its management is done decentralized.

This sounds akin to having a `Pool` managed by a group of governors and not a single entity? Perhaps, even instead of having a single owner have multiple ones, that would have weighted voting rights based on the amount of Pool share they hold?

> If we want to use Pool in both cases I propose renaming it to something like Managed Single Asset Tokenized Pool to show specifically that this pool is managed by someone

This is a great name, I'll update the spec. 


I understand you're using Flash Loans as an example, though, in your opinion, what benefit would flash loans bring? Also, do you see other uses cases for a unowned Pool?

**Reply by rothexD** [reply]
Yes Flash-Loan-Pool would be a type of public `Pool` that would have weighted fee voting shares. I just used it as an example because it was the first hing that came to mind, will have to think about how this applies to protocol based amms 

Another shared governance scenario that I was thinking about for a longer time would be a type of `Pool` where users can lock up assets up to a specific date and after that date if at least 50% of LP-token holders and 50% of LP-tokens agree the funds will be released to a beneficiary. If a quorum is not reached until an end date the assets can be reclaimed by their original holders. This would allow some type of shared community investment/funding (I am not sure about the legal side of this). You could then also do some type of release rounds where x% of assets get released. I have done no research yet if the community would even be interested in something like that. For example a community could fund the development of a video game using that and their next round of fund/ will only be released if milestones are reached and agreed on in a quorum.

I think a lot of what i think about lately can be solved using hooks but not sure if those ever come to main :D

**Reply by rothexD** [reply]
I think flash loans would only be useful if aams have experienced single side deposits and you can then use this imbalance to balance out all of the aams that have related pairs.

For example:

XRP-Flash-Loan-Pool

XRP/USDX
USDX/KUI
KUI/XRP

For example a lot of KUI are single side deposited into the USDX/KUI pair, you could then flash-loan and get XRP to trade for USDX and put those USDX back into the USDX/KUI pair while taking out KUI to swap it back to XRP. 

I hope I didn't miss something there... :D

**Reply by xVet** [reply]
> Thanks @xVet !
> 
> 
> 
> The idea is that a `PoolDelegate` is the pool's owner, and thus, the `PoolDelegateID` is not changeable. However, now that I think about it, there is no "delegacy" in this case. 
> 
> 
> 
> Do you think it would be better to have a separate `PoolOwner` and `PoolDelegate` where the `PoolOwner` has all permissions of the Delegate, including deleting the pool and changing the PoolDelegate. Whereas the PoolDelegate may only perform protocol specific operations? 
> 
> 

Exactly! I think it makes sense to change the name to `PoolOwner` and give the owner `AccountID` the ability to do a `AccountSet Flag` `asfPoolDelegate` for delegating functionality to a delegate.

Thoughts on that ?

**Reply by mvadari** [reply]
> I think it makes sense to change the name to `PoolOwner` and give the owner `AccountID` the ability to do a `AccountSet Flag` `asfPoolDelegate` for delegating functionality to a delegate.

I don't quite understand the need for the flag. Do you mean something similar to `NFTokenMinter`? If so, IMO in this case that should be on the pool and not on the account.

**Reply by Tapanito** [reply]
> Exactly! I think it makes sense to change the name to PoolOwner and give the owner AccountID the ability to do an AccountSet Flag asfPoolDelegate for delegating functionality to a delegate.

If we introduce a second entity, `PoolOwner,` I agree with 1). However, I'm not sure about the `asfPoolDelegate` flag. I think the `Pool` should be self-contained. The spec should not introduce `Pool` specific changes to other ledger entries. That is primarily to minimize coupling with other ledger entries. By minimizing coupling, we make the ledger entry easier to change. Or that is the hope at least.

That said, we can achieve the same functionality by introducing two fields, `PoolDelegate` and `PoolOwner` to the `Pool` object. What do you think about that?

**Comment by mvadari** [comment]
Can the `Preview` RPC be replaced with [XLS-69d](https://github.com/XRPLF/XRPL-Standards/discussions/199)?

**Comment by Tapanito** [comment]
Closing the discussion as the specification is in a good spot. The pull request can be found here: https://github.com/XRPLF/XRPL-Standards/pull/239
